convert2rhel/subscription.py

```diff
@@ -277,28 +338,125 @@ def args(self):
-            # and activation_key to use a file-based approach to passing the secrets.
-            output, ret_code = utils.run_subprocess([self.cmd] + self.args, print_cmd=False)
+        connection_opts = {}
+
+        if self.rhsm_hostname:
+            connection_opts["host"] = self.rhsm_hostname
+
+        if self.rhsm_port is not None:
+            connection_opts["port"] = self.rhsm_port
+
+        if self.rhsm_prefix:
+            connection_opts["handler"] = self.rhsm_prefix
+
+        connection_opts = dbus.Dictionary(connection_opts, signature="ss")
+        return connection_opts
 
-        return output, ret_code
+    def __call__(self):
+        """
+        Use dbus to register the system with subscription-manager.
+
+        Status of dbus on various platforms:
+            * RHEL6:
+                * DBUS-1.2.24 is present but may not be installed.
+                * Install the dbus package and run /etc/rc.d/init.d/messagebus start
+                * dbus-python-0.83.0 is available
+            * RHEL7:
+                * dbus-1.10.24 is installed and run by default
+                * dbus-python-1.1.9 is available
+            * RHEL8 & RHEL9
+                * dbus-1.12.x is installed and run by default
+                * python3-dbus-1.2.x is available
+
+        .. seealso::
+            Documentation for the subscription-manager dbus API:
+            https://www.candlepinproject.org/docs/subscription-manager/dbus_objects.html
+        """
+        # Note: dbus doesn't understand empty python dicts. Use dbus.Dictionary({}, signature="ss")
+        # if we need one in the future.
+        REGISTER_OPTS_DICT = dbus.Dictionary({"force": True}, signature="ss")
+
+        loggerinst.debug("Getting a handle to the system dbus")
+        system_bus = dbus.SystemBus()
+
+        # Create a new bus so we can talk to rhsm privately (For security:
+        # talking on the system bus might be eavesdropped in certain scenarios)
+        loggerinst.debug("Getting a subscription-manager RegisterServer object from dbus")
+        register_server = system_bus.get_object("com.redhat.RHSM1", "/com/redhat/RHSM1/RegisterServer")
+        loggerinst.debug("Starting a private DBus to talk to subscription-manager")
+        address = register_server.Start(
+            i18n.SUBSCRIPTION_MANAGER_LOCALE, dbus_interface="com.redhat.RHSM1.RegisterServer"
+        )
+
+        try:
+            # Use the private bus to register the machine
+            loggerinst.debug("Connecting to the private DBus")
+            private_bus = dbus.connection.Connection(address)
+            loggerinst.debug("Getting a subscription-manager Register object from dbus")
+            register_object = private_bus.get_object("com.redhat.RHSM1", "/com/redhat/RHSM1/Register")
+
+            if self.password:
+                loggerinst.debug("Registering via username/password: %s" % " ".join(hide_secrets(self.args)))
```

Comment:
A minor thing here: 

I feel that this should be a `info` instead of a `debug` to actually inform the user that we are registering the system. What you think?

Reply:
I don't think it's necessary. This is what they see now.

``` subscription-manager registration info ... 
09:45:31                 out:     ... activation key not found, using given username and password
09:45:31                 out:     ... username detected
09:45:31                 out:     ... password detected
```

If the credentials are incorrect, they'll get the `"Invalid username or password."` error. Then they can look at the convert2rhel.log file to see what username they used or even better look at how they called convert2rhel so they can see even the password they used.

Reply:
Makes sense to me! Thanks